### PR TITLE
BUG, ENH: fix array2string rounding bug by adding min_digits option

### LINKED ISSUE #18609: floatmode=unique sometimes giving wrong rounding
Let `a` be a special float64:
```python
In [26]: a = np.float64.fromhex('-1p-97')

In [27]: a
Out[27]: -6.310887241768095e-30
```

It has `uniquestra`, a shortest decimal string that uniquely identifies it (thus can be eval-ed to reproduce it exactly):
```python
In [33]: uniquestra = np.format_float_scientific(a, unique=True)

In [34]: uniquestra
Out[34]: '-6.310887241768095e-30'

In [35]: eval(uniquestra)==a
Out[35]: True
```

Because `a` is a special float, unlike most other floats, it has a correctly-rounded precision=15 decimal string _that is different from `uniquestra`_:
```python
In [36]: roundedstra = np.format_float_scientific(a, unique=False, precision=15)

In [37]: roundedstra
Out[37]: '-6.310887241768094e-30'

In [38]: eval(roundedstra)==a
Out[38]: False
```
So far, so good. This kind of thing is why `format_float_scientific` has the `unique` flag, to choose between these two cases depending on which is needed in a given situation.

The problem comes with `array2string`, who takes a `floatmode=unique` option that should give `uniquestr` and instead gives `roundedstr`. The problem is that internally `array2string` uses `FloatingFormat` which in `floatmode='unique'` mode effectively uses format_float_scientific twice: Once with `unique=True, precision=None` to figure out how many digits of precision are enough to uniquely represent the float, and then a second time with `unique=False, precision=foo` to actually do the formatting. Most of the time this correctly gives a unique representation of the float, but for a special float like `a` it fails and gives `roundedstra` instead of `uniquestra`. 

### Reproducing code example:

```python
import numpy as np
a = np.float64.fromhex('-1p-97')
assert eval(np.array2string(a, floatmode='unique'))==a
```

### NumPy/Python version information:
```
1.20.1 3.8.8 (default, Feb 24 2021, 13:46:16) 
[Clang 10.0.0 ]
```

### PR DESCRIPTION
**Pull request by ahaldane**
Fixes #18609

This PR adds a new `min_digits` argument to the dragon4 float printing methods, such as `np.format_float_positional`. This kwd guarantees that at least the given number of digits will be printed, when printing in `unique=True` mode, even if the extra digits are unnecessary to uniquely specify the value. It is the counterpart to the `precision` argument which sets the maximum number of digits to be printed. (When `unique=False` fixed-precision mode, it has no effect, and the `precision` arg fixes the number of digits)

Using this new keyword, I then fix a bug in arrayprint ( #18609). The bug in the old code is that, in order to guarantee that all elements of an array print using the same number of digits, we turned off `unique=True` because we needed to use fixed-precision mode to fix the number of digits. However, unique mode rounds differently from fixed-precision mode, so the rounding was wrong. This PR instead now stays in unique mode to print the elements, but uses the `min_digits` and `precision` arguments to fix the number of output digits.

It looks like a lof of chages, but a lot of it is just introducing the `min_digits` arg into all the function signatures in dragon4 and in the user api.

### CODE REVIEW

**Review by BvB93 — COMMENTED** [review]
Can you update the annotations in the `arrayprint.pyi` stub file?
By the looks of it the type of `min_digits` should be `Optional[int]`.

https://github.com/numpy/numpy/blob/ee9c8f8a3a3385939a0b8a4bc716693dd2995e15/numpy/core/arrayprint.pyi#L98-L116

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -1027,11 +1030,12 @@ def format_float_scientific(x, precision=None, unique=True, trim='k',
         If `True`, use a digit-generation strategy which gives the shortest
         representation which uniquely identifies the floating-point number from
         other values of the same type, by judicious rounding. If `precision`
-        was omitted, print all necessary digits, otherwise digit generation is
-        cut off after `precision` digits and the remaining value is rounded.
+        is given fewer digits than necessary can be printed, or if `min_digits`
```
Comment (seberg) [inline numpy/core/arrayprint.py:1033]:
```suggestion
        is given fewer digits than necessary can be printed. If `min_digits`
```

**Review by seberg — APPROVED** [review]
Thanks Allan. I took the liberty to rebase, if we want a backport give me a ping and I will do that.

I won't claim the have followed all the logic in `dragon4.c` but the code changes look all good and the tests sufficiently thorough.

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -1048,6 +1052,10 @@ def format_float_scientific(x, precision=None, unique=True, trim='k',
     exp_digits : non-negative integer, optional
         Pad the exponent with zeros until it contains at least this many digits.
         If omitted, the exponent will be at least 2 digits.
+    min_digits : non-negative integer or None, optional
+        Minimum number of digits to print. Only has an effect if `unique=True`
+        in which case additional digits past those necessary to uniquely
+        identify the value may be printed, rounding the last additional digit.
```
Comment (seberg) [inline numpy/core/arrayprint.py:1058]: Hmm, an intended comment got lost. Was thinking we could try to simplify this a bit:
```suggestion
        Minimum number of digits to print. This only has an effect for `unique=True`.
        In that case more digits than necessary to uniquely identify the value may
        be printed. The last additional digit is rounded unbiased.
```
but happy to just leave it.

**Comment by charris** [comment]
Could use a release note.

File: numpy/core/arrayprint.py [hunk 3]
```diff
@@ -1048,6 +1052,11 @@ def format_float_scientific(x, precision=None, unique=True, trim='k',
     exp_digits : non-negative integer, optional
         Pad the exponent with zeros until it contains at least this many digits.
         If omitted, the exponent will be at least 2 digits.
+    min_digits : non-negative integer or None, optional
+        Minimum number of digits to print. This only has an effect for
+        `unique=True`. In that case more digits than necessary to uniquely
+        identify the value may be printed and rounded unbiased.
+
```
Comment (charris) [inline numpy/core/arrayprint.py:1059]: Needs `.. versionadded:: 1.21.0`.

File: numpy/core/arrayprint.py [hunk 4]
```diff
@@ -1123,6 +1138,10 @@ def format_float_positional(x, precision=None, unique=True,
     pad_right : non-negative integer, optional
         Pad the right side of the string with whitespace until at least that
         many characters are to the right of the decimal point.
+    min_digits : non-negative integer or None, optional
+        Minimum number of digits to print. Only has an effect if `unique=True`
+        in which case additional digits past those necessary to uniquely
+        identify the value may be printed, rounding the last additional digit.
```
Comment (charris) [inline numpy/core/arrayprint.py:1144]: Needs `.. versionadded:: 1.21.0`.

**Comment by charris** [comment]
Thanks Allan

**Pull request merged**